worlds/tloz/Rules.py

```diff
@@ -0,0 +1,145 @@
+    # No requiring anything in a shop until we can farm for money
+    for location in shop_locations:
+        add_rule(world.get_location(location, player),
+                 lambda state: state.has_group("weapons", player))
+
+    # Everything from 4 on up has dark rooms
+    for level in tloz_world.levels[4:]:
+        for location in level.locations:
+            add_rule(world.get_location(location.name, player),
+                     lambda state: state.has_group("candles", player)
+                                   or (state.has("Magical Rod", player) and state.has("Book", player)))
+
+    # Everything from 5 on up has gaps
+    for level in tloz_world.levels[5:]:
+        for location in level.locations:
+            add_rule(world.get_location(location.name, player),
+                     lambda state: state.has("Stepladder", player))
+
+    add_rule(world.get_location("Level 5 Boss", player),
+             lambda state: state.has("Recorder", player))
+
+    add_rule(world.get_location("Level 6 Boss", player),
+             lambda state: state.has("Bow", player) and state.has_group("arrows", player))
+
+    add_rule(world.get_location("Level 7 Item (Red Candle)", player),
+             lambda state: state.has("Recorder", player))
+    add_rule(world.get_location("Level 7 Boss", player),
+             lambda state: state.has("Recorder", player))
+    if world.ExpandedPool[player]:
+        add_rule(world.get_location("Level 7 Key Drop (Stalfos)", player),
+                 lambda state: state.has("Recorder", player))
+        add_rule(world.get_location("Level 7 Bomb Drop (Digdogger)", player),
+                 lambda state: state.has("Recorder", player))
+        add_rule(world.get_location("Level 7 Rupee Drop (Dodongos)", player),
+                 lambda state: state.has("Recorder", player))
+
+    for location in food_locations:
+        if world.ExpandedPool[player] or "Drop" not in location:
+            add_rule(world.get_location(location, player),
+                     lambda state: state.has("Food", player))
+
+    add_rule(world.get_location("Level 8 Item (Magical Key)", player),
+             lambda state: state.has("Bow", player) and state.has_group("arrows", player))
+    if world.ExpandedPool[player]:
+        add_rule(world.get_location("Level 8 Bomb Drop (Darknuts North)", player),
+                 lambda state: state.has("Bow", player) and state.has_group("arrows", player))
+
+    for location in tloz_world.levels[9].locations:
+        add_rule(world.get_location(location.name, player),
+                 lambda state: state.has("Triforce Fragment", player, 8) and
+                               state.has_group("swords", player))
+
+    # Yes we are looping this range again for Triforce locations. No I can't add it to the boss event loop
+    for level in range(1, 9):
+        add_rule(world.get_location(f"Level {level} Triforce", player),
+                 lambda state: state.has(f"Boss {level} Defeated", player))
```

Comment:
Creating a lambda in a loop like this causes all these lambdas to end up referenced with `level` as it is at the end of the loop, which causes them all to logically require Boss 9 Defeated. This would be a way to do this correctly:
```
        add_rule(world.get_location(f"Level {level} Triforce", player),
                 lambda state, l=level: state.has(f"Boss {l} Defeated", player))
```